Camera Overhaul Core

DragonMoffon opened the pull request:
This PR includes the core functionality in a stable (without unit tests or docs) state. This includes a stripping out of all references to the old system. 

Removed:
- the set_viewport method. Has been replaced by a variety of options. Either directly setting matrices, using the default camera, or using a Camera2D
- The old Camera class. This will not be replaced in a backward-compatible manner as it is deeply flawed and has been replaced by Camera2D

Removed but coming in the future:
- Screen shake.

There are a few issues with the current camera system in arcade. Mainly it only supports a 2D orthographic camera with limited zoom and rotation functionality. This refactor aims to create a new system that allows users to easily make fully 3D orthographic or perspective cameras which are easy to use and create custom functionality for.

The refactor aims to be mostly backwards-compatible with the previously implemented SImpleCamera and Camera.

The refactor includes:
A ViewData PoD that holds all of the information needed to make a view matrix and set the viewport.
Two ProjectionData PoDs hold the data for making either an orthographic or projection matrix.

The Projection Protocol to link the ProjectionData PoDs together (may be removed as it is unnecessary)
The Projector Protocol defines any object that provides the Use, Activate, and get_map_coordinates (may be renamed) methods.
The Camera Protocol defines an object which holds both a ViewData PoD and a ProjectionProtocol Object.

The OrthographicCamera class uses the ViewData and OrthographicProjectionData PoDs to provide an easy way to create and use an orthographic projection matrix, but it alone does not provide ways of controlling the camera.
The PersepctiveCamera class is the same as the OrthographicCamera class, except it uses the PerspectiveProjectionData PoD.

It provides a backwards-compatible SimpleCamera class implementation with the same methods.
The Camera2D class provides a simple interface to control nearly every camera aspect and will be what 99% of arcade users will need. It does not provide animations or any other similar logic but provides easy properties and methods for setting the different variables that are used by the camera. The only clamps on control are lack of access to the forward vector and only accepting 2D vectors.

The DefaultProjector class is provided to have a base camera only to be used internally by Arcade. This class is designed to be used as the default "current_camera". This is because of a significant bug currently plaguing Arcade. If the user has never used a Camera and they try to use a camera as a context manager, the current camera will be equal to None and cause a crash.

See Issue https://github.com/pythonarcade/arcade/issues/1853 for more details

DragonMoffon committed:
Setup primary files for experimental refactor of the Camera, Backgrounds, and Scenes.

See the
#Possible Scene Improvements
#Cameras
topics in the arcade-dev forum on the Python Arcade discord server.

DragonMoffon committed:
Merge branch 'pythonarcade:development' into scene-refactor-experiment

DragonMoffon committed:
Completed the Camera2DOrthographic class with doc strings and comments, and started the Camera2DController, and SimpleCamera classes.

See the
#Possible Scene Improvements
#Cameras
topics in the arcade-dev forum on the Python Arcade discord server for more info.

DragonMoffon committed:
Slight change to Cameras to allow for a perspective projection matrix. Created basic Perspective and Orthographic Cameras.

See the
#Possible Scene Improvements
#Cameras
topics in the arcade-dev forum on the Python Arcade discord server for more info.

DragonMoffon committed:
Completed Orthographic Camera. Had some issues with the view matrix, This has been fixed and applied to both of the base cameras. Added a new get map coordinates function (open to change). Placed framework for backwards compatible simple camera.

See the
#Possible Scene Improvements
#Cameras
topics in the arcade-dev forum on the Python Arcade discord server for more info.

DragonMoffon committed:
Update camera_refactor.py

Finished Simple Camera. Is backwards compatible with current Simple Camera implementation.

DragonMoffon committed:
PR cleanup

Cleaning up PR to only include camera refactor

DragonMoffon committed:
New Camera Code Integration

Moved experimental code into new "cinematic" folder within arcade.

Also made the default camera in arcade the "DefaultCamera" class. and made it's type be "Projector"

DragonMoffon committed:
Code inspection Clean up

DragonMoffon committed:
Inspection Fix 2

DragonMoffon committed:
Code inspection fix 3

DragonMoffon committed:
Round 4

DragonMoffon committed:
Writitng initial Unit Tests

Created files for unit tests, and wrote a few.

Started work on Camera2D (replacement for simple camera)

DragonMoffon committed:
Merge branch 'pythonarcade:development' into scene-refactor-experiment

DragonMoffon committed:
Finished base of Camera2D class.

The basics for Camera2D have been provided with full doc strings. Other helper methods may be added in the future.

DragonMoffon committed:
Merge branch 'scene-refactor-experiment' of https://github.com/DragonMoffon/arcade into scene-refactor-experiment

DragonMoffon committed:
code-inspection clean-up on Camera2D

Fixed `mypy`, `pyright`, `ruff` errors.

Also added __all__ property to every file
for a better importing experience.

NOTE arcade/camera.py is still there, and it does not match the current system so the code-inspection still complains. Will resolve later.

DragonMoffon committed:
Removed all reference to old camera system.

This included deleting `arcade/camera.py`, and fixing the ui and sections to use either the Default Ortho Projector.

NOTE I removed a quick index to the `camera.rst`. That will need to be fixed.

Hey look I linted before pushing for once!

DragonMoffon committed:
whoops circular imports

DragonMoffon committed:
circular imports 2

DragonMoffon committed:
type checking

DragonMoffon committed:
Started work on some controllers

Made a few function controllers which are mega simple.

Also changed name from arcade.cinematic to arcade.camera.

Also moved the controllers to arcade.camera.controllers.

DragonMoffon committed:
fixing silly pycharm muck-up

when I changed the file name it didn't update any imports tsk tsk.

DragonMoffon committed:
removing doc ref

Have not setup camera documentation so removing old ref.

DO NOT PULL PR UNTIL FIXED.

DragonMoffon committed:
Updated Orthographic Unit Tests

grrrrr pycharm

DragonMoffon committed:
Fixed all the examples

NOTE this is a quick fix. It removed shaking from two examples. CANNOT BE PULLED IN WHILE THIS IS UNRESOLVED.

Weirdly the linters didn't pick up on these errors

DragonMoffon committed:
linting fix 1-million

DragonMoffon committed:
1-million and 1

DragonMoffon committed:
1-million and 2

DragonMoffon committed:
MOAR example fixing

DragonMoffon committed:
Setup 4 Splines for SplineController

setup lerp, quadratic, cubic, b-spline

Contemplating how to do spline controller

DragonMoffon committed:
Removed splines from this PR

Removed Splines
Made Isometric Controller
Fixed small issue with facing direction
doing temp rendering test so don't mind `log.png`

DragonMoffon committed:
Merge branch 'pythonarcade:development' into scene-refactor-experiment

DragonMoffon committed:
General work

DragonMoffon committed:
Merge branch 'scene-refactor-experiment' of https://github.com/DragonMoffon/arcade into camera-overhaul

DragonMoffon committed:
Cleaned up Offscreen renderer

Added 'nother default glsl shader. Also cleaned up some linting.

DragonMoffon committed:
Merge branch 'pythonarcade:development' into camera-overhaul

DragonMoffon committed:
Squashed commit of the following:

commit 7d56d39f6dfdeea48be659b4871ced059fe2e91e
Author: DragonMoffon <happydog.aja@gmail.com>
Date:   Fri Aug 18 05:16:55 2023 +1200

    Created a camera shake class

    Created a camera shake class. It isn't exactly like the shake provided before so people might what to have a review.

DragonMoffon committed:
added screen shake doc

DragonMoffon committed:
lovely useless doc strings

NOTE: The doc strings aren't useless. The code just is. Going to completely change the class.

DragonMoffon committed:
Merge remote-tracking branch 'upstream/development' into camera-overhaul

DragonMoffon committed:
improving screen shake

DragonMoffon committed:
touch-ups

DragonMoffon committed:
Finalised Core camera functionality

- Removed all reference to arcade.ser_viewport method

- fixed incorrect forward direction for SimpleCamera

- Includes projectors, SimpleCamera, and Camera2D

- Includes minimal functions for moving cameras.

DragonMoffon committed:
Merge branch 'pythonarcade:development' into Camera-Overhaul-Core

DragonMoffon committed:
Removed references to projection_2d

alejcas commented:
Quick question: Does this new camera work with arcade collision methods when applying zoom or rotation?

DragonMoffon commented:
@janscas Generally they should as the collisions are done in world space which the cameras don't touch. The cameras only affect where it looks like sprites are. Should still be double-checked, but I would be surprised.

DragonMoffon committed:
Merge branch 'Camera-Overhaul-Core' of https://github.com/DragonMoffon/arcade into Camera-Overhaul-Core

DragonMoffon committed:
Fixing unit tests

DragonMoffon committed:
Merge branch 'pythonarcade:development' into Camera-Overhaul-Core

DragonMoffon committed:
Solved pytest issues in GUI due to Camera Code

DragonMoffon committed:
Fixed small linting issue

DragonMoffon committed:
Fixed double brackets caused by find and replace

DragonMoffon committed:
Whoops didn't run integration tests

DragonMoffon committed:
Created Camera Shake Controller.

Created a camera shake controller to replace the one removed.

DragonMoffon committed:
Completed Doc Strings

DragonMoffon committed:
Replacing all references of SimpleCamera with Camera2D

DragonMoffon committed:
Fixing integration test errors, and examples.

Did not get to all examples

DragonMoffon committed:
attempting to fix gui part 1

DragonMoffon committed:
Fixed erroneous ui behaviour with new camera.

DragonMoffon committed:
fixed linting issues

pvcraven commented on doc/tutorials/lights/01_light_demo.py:
```diff
@@ -25,9 +25,8 @@ def __init__(self, width, height, title):
-        self.view_left = 0
-        self.view_bottom = 0
+        # camera for scrolling
+        self.cam = None
```
I'd prefer to spell out camera.

pvcraven commented on tests/unit/camera/test_default_camera.py:
Are these placeholder files? Still looking for tests? If so, probably need some comments stating that.

DragonMoffon replied:
Yup, tests are still not done.

pvcraven commented on arcade/camera/simple_camera.py:
```diff
@@ -0,0 +1,381 @@
+                 projection_data: Optional[OrthographicProjectionData] = None
+                 ):
+        """
+        Initialise a Simple Camera Instance with either Camera PoDs or individual arguments
```
"Initialize" for u.s. spelling.

pvcraven commented on arcade/examples/platform_tutorial/06_camera.py:
```diff
@@ -121,19 +121,17 @@ def on_key_release(self, key, modifiers):
         player_centered = screen_center_x, screen_center_y
 
-        self.camera.move_to(player_centered)
+        arcade.camera.controllers.simple_follow_2D(0.2, player_centered, self.camera.view_data)
```
Probably should use a named parameter or variable so that 0.2 isn't a random number appearing out of no where. I think elsewhere you used a camera_panning variable of some sort.

DragonMoffon committed:
Merge branch 'pythonarcade:development' into Camera-Overhaul-Core

DragonMoffon committed:
Fixing Issues from pvcraven´s review

resolved pvcraven´s review. Also created a perspective camera demo. setup unit tests.

DragonMoffon committed:
Merge branch 'Camera-Overhaul-Core' of https://github.com/DragonMoffon/arcade into Camera-Overhaul-Core

MiCurry commented on arcade/camera/camera_2d.py:
```diff
@@ -0,0 +1,713 @@
+            projection: A 4-float tuple which defines the world space
+                        bounds which the camera projects to the viewport.
+            near: The near clipping plane of the camera.
+            far: The far clipping place of the camera.
```
Should `place` be `plane` here?

DragonMoffon replied:
plane

MiCurry commented on arcade/camera/camera_2d.py:
```diff
@@ -0,0 +1,713 @@
+class Camera2D:
+    """
+    A simple orthographic camera. Similar to SimpleCamera, but takes better advantage of the new data structures.
+    As the Simple Camera is depreciated any new project should use this camera instead.
```
There should be a comma after 'deprecated':

> As the Simple Camera is depreciated, any new project should use this camera instead.

MiCurry commented on arcade/camera/camera_2d.py:
```diff
@@ -0,0 +1,713 @@
+    def view_data(self) -> CameraData:
+        """
+        Return the view data for the camera. This includes the
+        viewport, position, forward vector, up direction, and zoom.
```
I think `CameraData` just contains `position`, `up`, `forward` and `zoom` and **not** the `viewport`, correct?

MiCurry replied:
A print of the view_data property shows:
```
CameraData(position=(0.0, 0.0, 0.0), up=(0.0, 1.0, 0.0), forward=(0.0, 0.0, -1.0), zoom=1.0)
```

DragonMoffon replied:
yup this doc string was written before a change in the data structures good call

DragonMoffon committed:
Fixed issues pointed out by MiCurry

thanks MiCurry for the editing on that!

Also did linting. Currently fails on the camera_perspective_demo.py because of how I threw it together. Will fix in next commit

DragonMoffon committed:
Improved Camera Shake and Unit Tests

Changed the screen shake to store the
shake dir but not vector so it should work with rotating the camera now.

Added a `readjust_camera` method to the SceenShakeController.

Also finished it´s unit tests.

DragonMoffon committed:
Improved Orthographic and Perspective Unit Tests

DragonMoffon committed:
Added cull_mode to spritelists

Currently only picks between 'orthographic' and 'disabled'

DragonMoffon committed:
Started working on simple camera controller unit tests

MiCurry commented on arcade/camera/camera_2d.py:
```diff
@@ -0,0 +1,713 @@
+        The height of the viewport.
+        Defines the number of pixels drawn too vertically.
+        """
+        return self._projection.viewport[3]
```
Same thing as above with the viewport height. This will fail if the viewport starts at anything other than the origin. I think it should be:

`return self._projection.viewport[3] - self._projection.viewport[1]`

DragonMoffon replied:
The viewport is stored as left, bottom, width, and height. So this is not necessary.

MiCurry commented on arcade/camera/camera_2d.py:
```diff
@@ -0,0 +1,713 @@
+        The width of the viewport.
+        Defines the number of pixels drawn too horizontally.
+        """
+        return self._projection.viewport[2]
```
I think this is the viewport width only when the viewport starts at (0,0) (e.g. (0, 0, x, y), but this will produce an incorrect width when the viewport starts at anything other than the origin. Should it be something like:

`return self._projection.viewport[2] - self._projection.viewport[0]`

DragonMoffon replied:
See the previous comment with the viewport being stored as left, bottom, width, and height.

DragonMoffon committed:
continue perspective work

DragonMoffon commented:
Closed as a new PR was made to replace it. Third times a charm :)
